crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+        let mut finished: CheckpointExecutionBuffer = FuturesOrdered::new();
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
+
+        loop {
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((seq_num, checkpoint, next_committee))) = finished.next() => {
+                    match next_committee {
+                        None => {
+                            self.metrics.last_executed_checkpoint.set(seq_num as i64);
+                            self.checkpoint_store
+                                .update_highest_executed_checkpoint(&checkpoint)
+                                .unwrap();
+
+                            let highest_synced = self
+                                .checkpoint_store
+                                .get_highest_synced_checkpoint_seq_number().unwrap()
+                                .unwrap_or(0);
+                            // If we got here, `highest_scheduled_checkpoint` must have been set,
+                            // so safe to unwrap
+                            if self.highest_scheduled_checkpoint.unwrap() < highest_synced && !self.end_of_epoch {
+                                self.schedule_checkpoint(seq_num.saturating_add(1), &mut finished).await;
+                            }
+                        }
+                        // Last checkpoint of epoch -- Note that we must not update the
+                        // highest executed watermark until after reconfig has completed!
+                        Some(committee) => return Some((checkpoint, committee)),
+                    }
+                }
+                // Read from StateSync channel if we have capacity to schedule more checkpoints.
+                // If we're at end of epoch, skip this branch so that we can effectively await
+                // execution of all remaining checkpoints of the epoch (which should be scheduled)
+                // and call reconfig.
+                received = self.mailbox.recv(), if finished.len() < task_limit && !self.end_of_epoch => match received {
+                    Ok(checkpoint) => {
+                        let last_to_exec = checkpoint.sequence_number() as u64;
+                        let highest_executed_checkpoint = self.checkpoint_store.get_highest_executed_checkpoint_seq_number().unwrap_or_else(
+                            |err| {
+                                panic!("Failed to read highest executed checkpoint sequence number: {:?}", err)
+                            }
+                        );
+
+                        // Note that either of these can be higher. If the node crashes with many
```

Comment:
Is crash recovery is the only case when they would be inconsistent, why not simply recover `self.highest_scheduled_checkpoint` from the persisted db when initializing the executor?

Reply:
That is not really a watermark. It's just some in-memory state that I keep in order to ensure that I do not re-schedule the same checkpoint sequence number that is currently pending execution (which would happen if I only rely on the `highest_executed_checkpoint` watermark)